PYME/recipes/kinetics_recipes.py

```diff
@@ -0,0 +1,428 @@
+    offTimesColName = CStr('off_times')
+
+    offTimesBinSizeInSeconds = Float(2.5)
+
+    minOffTimeSecondsToFit = Float(0.0)
+    maxOffTimeSecondsToFit = Float(5000.0)
+
+    off_init_condition_multiple = Int(1)
+
+    exposure_time = Float(-1.)
+
+    def execute(self, namespace):
+        from PYME.Analysis.BleachProfile.kinModels import eimod
+
+        #todo: add higher order fits
+
+        oncol = self.onTimesColName
+        offcol = self.offTimesColName
+
+        inp = namespace[self.inputName]
+
+        if 'mdh' in dir(inp):
+            mdh = inp.mdh
+            cyctime = mdh.getEntry('Camera.CycleTime')
+        else:
+            cyctime = self.exposure_time
+
+        ds = tabular.mappingFilter(inp)
+
+        off_fit_limit_index = np.round(self.maxOffTimeSecondsToFit/cyctime).astype('int')
+
+        offtimeslimit = np.round(self.minOffTimeSecondsToFit/cyctime).astype('int')
+
+        onbins = np.arange(1, max(ds[oncol])+1, 1)
+        offbins = np.arange(offtimeslimit, off_fit_limit_index, np.round(self.offTimesBinSizeInSeconds/cyctime).astype('int'))
+
+
+        n_on, bins_on = np.histogram(ds[oncol], bins=onbins)
+        n_off, bins_off = np.histogram(ds[offcol], bins=offbins)
+
+        bins_on = bins_on[:-1]
+        bins_off = bins_off[:-1]
+
+        res_off = FitModelWeighted(eimod, [n_off[0]*self.off_init_condition_multiple, 1], n_off, 1. / (np.sqrt(n_off) + 1), bins_off * cyctime)
+
+        res_on = FitModelWeighted(eimod, [n_on[1], 1], n_on[1:], 1. / (np.sqrt(n_on[1:]) + 1), bins_on[1:] * cyctime)
+
+        pylab.figure()
+
+        pylab.bar(bins_on*cyctime, n_on, width=np.diff(bins_on)[0]*cyctime, alpha=0.4, fc=colours[0])
+        pylab.plot(np.linspace(bins_on[0], max(bins_on), max(bins_on) * 10) * cyctime,
+                   eimod(res_on[0], np.linspace(bins_on[0], max(bins_on), max(bins_on) * 10) * cyctime), colours[0],
+                   lw=3)
+
+        pylab.ylabel('Events')
+        pylab.xlabel('Event Duration [s]')
+        pylab.xlim((0, np.round(n_on[-1]*1.2).astype('int')))
+        pylab.title('Blink Fluorescence Durations')
+        pylab.figtext(.6, .8 - .05, 'tau = %3.4f' % (res_on[0][1]), size=18, color=colours[1])
+        pylab.show()
+
+
+        pylab.figure()
+
+        pylab.xlim(right=np.round(self.maxOffTimeSecondsToFit/cyctime).astype('int'))
+        pylab.bar(bins_off * cyctime, n_off, width=np.diff(bins_off)[0] * cyctime, alpha=0.3, fc=colours[2])
+        pylab.plot(np.linspace(bins_off[0], max(bins_off), max(bins_off) * 10) * cyctime,
+                   eimod(res_off[0], np.linspace(1, max(bins_off), max(bins_off) * 10) * cyctime),
+                   colours[3], lw=3, alpha=.8)
+        pylab.ylabel('Events')
+        pylab.xlabel('Event Duration [s]')
+        pylab.ylim((.9, pylab.ylim()[1]))
+        pylab.title('Blink dark times')
+
+        pylab.figtext(.6, .8 - .05, 'tau = %3.4f' % (res_off[0][1]), size=18, color=colours[1])
+
+        pylab.show()
```

Comment:
You don't want to be doing any gui stuff in recipe modules. See report outputs in recipes/outputs.py, the idea being you can use jinja templating to pop plots into htmls.

Reply:
Should mention an alternative to the reports, since that definitely can take some time to get formatted as you like, etc., is just to move all of this plotting code into the vis gui plugin, so it only gets called when you are already doing gui code. Reports would be 'cooler' and maybe a bit cleaner, but if these are just for sanity checks as you run a recipe and you are unlikely to batch loads of them, visgui pyplot stuff is probably the winner of the cost-benefit.

Reply:
The sanity check aspect would be nice, but in the end I definitely want to add the reports, so I'll probably end up doing both